asab/library/providers/git.py

```diff
@@ -0,0 +1,158 @@
+
+L = logging.getLogger(__name__)
+
+#
+
+try:
+	import pygit2
+except ImportError:
+	L.critical("Please install pygit2 package to enable Git Library Provider. >>> pip install pygit2")
+	raise SystemExit("Application exiting... .")
+
+
+class GitLibraryProvider(FileSystemLibraryProvider):
+	"""
+	Read-only git provider to read from remote repository.
+	It clones a remote git repository to a temporary directory and then uses the
+	FileSystemLibraryProvider to read the files.
+	To read from local git repository, please use FileSystemProvider.
+
+	Configuration:
+	(Use either deploytoken, publickey+privatekey for SSH option, or username and password and HTTP access.)
+
+	```
+	[library]
+	providers=git+<URL or deploy token>#<branch name>
+
+	[library:git]
+	publickey=<absolute path to file>
+	privatekey=<absolute path to file>
+	username=johnsmith
+	password=secretpassword
+	repodir=<optional location of the repository cache>
+	```
+	"""
+	def __init__(self, library, path):
+
+		# The branch can be optionally specified in the URL fragment (after '#')
+		split_path = path.split("#", 1)
+		if len(split_path) > 1:
+			self.Branch = split_path[-1]
+			self.URL = split_path[0][4:]
+		else:
+			self.Branch = None
+			self.URL = path[4:]
+
+		self.Callbacks = pygit2.RemoteCallbacks(get_git_credentials(self.URL))
+
+		repodir = Config.get("library:git", "repodir", fallback=None)
+		if repodir is not None:
+			self.RepoPath = os.path.abspath(repodir)
+		else:
+			tempdir = tempfile.gettempdir()
+			self.RepoPath = os.path.join(
+				tempdir,
+				"asab.library.git",
+				hashlib.sha256(path.encode('utf-8')).hexdigest()
+			)
+
+		super().__init__(library, self.RepoPath, set_ready=False)
+
+		from ...proactor import Module
+		self.App.add_module(Module)
+		self.ProactorService = self.App.get_service("asab.ProactorService")
+		self.PullLock = False
+
+		self.App.TaskService.schedule(self.intialize_git_repo())
+		self.App.PubSub.subscribe("Application.tick/60!", self._periodic_pull)
+
+
+	async def _periodic_pull(self, event_name):
+		if self.PullLock:
+			return
+
+		self.PullLock = True
+
+		try:
+			await self.ProactorService.execute(self.pull)
+		finally:
+			self.PullLock = False
+
+
+	async def intialize_git_repo(self):
+
+		def init_task():
+			if pygit2.discover_repository(self.RepoPath) is None:
+				# For a new repository, clone the remote bit
+				os.makedirs(self.RepoPath, mode=0o700)
+				self.GitRepository = pygit2.clone_repository(self.URL, self.RepoPath, callbacks=self.Callbacks, checkout_branch=self.Branch)
+			else:
+				# For existing repository, pull the latest changes
+				self.GitRepository = pygit2.Repository(self.RepoPath)
+				self.pull()
+
+		await self.ProactorService.execute(init_task)
+
+		try:
+			assert self.GitRepository.remotes["origin"] is not None
+		except (KeyError, AssertionError):
+			L.error("Connection to remote git repository failed.")
+			# The library will not get ready ... maybe we can retry init_test() in a while
```

Comment:
Já si myslím, že většina problémů bude špatná konfigurace a merge conflicts. Což je něco, s čím opakující se inicializace nepomůže. Tohle je jenom jeden z bodů, kde to bude padat, pokud bude špatná konfigurace. Nejčastější podle mě bude, že se pygit2.clone_repository nezvládne checkoutnout do Branche, která neexistuje. Nicméně uživatel nebude vědět, proč mu to failnulo. Jsem přesvědčená, že může nastat situace, kdy to failne při prvním pokusu o merge, který se teď ale děje až v runtimu. Pokud je argument, že nová URL = nové repo, tak bych odebrala z konfigurace `repodir`.

Reply:
Spis bych se zde bal toho, ze kdyz nekdo bude startovat microservice a bude jiz mit naklonovane repo z minula ... a zaroven treba spadne sitova konektivita na git server, tak ta microservice musi stale nastartovat.

Ale toto muzeme dotahnout ex-post.

Reply:
tak to stejně spadne na prvním git fetch, který nepůjde udělat...

Reply:
tak to musi spadnout tak, aby to bezelo dal - tj. jen warning.

Kriticka chyba zde je pouze prvni clone - ten projit musi.